### PR TITLE
Adds start_link_supervised

### PR DESCRIPTION
**Pull request by joladev**
Adds a version of `start_supervised` and `start_supervised!` that explicitly links the started process to the test process, ensuring that any crashes are propagated to the test, failing it and providing a helpful test failure output.

More context on the change here: https://github.com/elixir-lang/elixir/issues/11737

### Open questions
1. Can we test `start_link_supervised` with a process that crashes where we assert on the EXIT message received without race conditions?
2. Are any other tests missing? I didn't want to repeat every `start_supervised` test since it just delegates to it under the hood, but the `stop_supervised` one was relevant
3. Do the docs make sense?

### CODE REVIEW

File: lib/ex_unit/lib/ex_unit/callbacks.ex [hunk 1]
```diff
@@ -475,6 +509,17 @@ defmodule ExUnit.Callbacks do
     end
   end
 
+  defp pid_for_child(sup, id) do
+    result =
+      sup
+      |> Supervisor.which_children()
+      |> Enum.find(fn {child_id, _pid, _type, _modules} -> child_id == id end)
```
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/callbacks.ex:516]: I think you can use List.keyfind :)

**Comment by josevalim** [comment]
I think we should only provide a `!` variant, as if the link crashes, the process will crash too. :)

File: lib/ex_unit/lib/ex_unit/callbacks.ex [hunk 2]
```diff
@@ -415,6 +415,21 @@ defmodule ExUnit.Callbacks do
     end
   end
 
+  @doc """
+  Same as `start_supervised/2` but also links the supervised process to the test process.
+  This means that if the process that was started crashes that crash is propagated to
```
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/callbacks.ex:420]:
```suggestion

  This means that if the process that was started crashes that crash is propagated to
```

File: lib/ex_unit/lib/ex_unit/callbacks.ex [hunk 3]
```diff
@@ -443,6 +447,22 @@ defmodule ExUnit.Callbacks do
 
   defp start_supervised_error(reason), do: Exception.format_exit({:start_spec, reason})
 
+  @doc """
+  Same as `start_link_supervised/2` but returns the PID on success and raises if
+  not started properly.
```
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/callbacks.ex:452]: The docs need to be updated now as the non-bang variant no longer exists. :)

Comment (joladev) [reply lib/ex_unit/lib/ex_unit/callbacks.ex]: Right!

**Review by josevalim — APPROVED** [review]

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**